src/ipss_cell.hpp

```diff
@@ -0,0 +1,201 @@
+        next_rates_it_++;
+        update_rates();
+
+        // Step through the rate_change vector if begin_step does not fall in the
+        // range of the current pointers
+        while (start_step_ >= next_rates_it_->first) {
+            // updates the current rates and dt
+            current_rate_it_++;
+            next_rates_it_++;
+            update_rates();
+        }
+
+        // Set the internal step counter to the start of the cell
+        current_step_ = start_step_;
+
+        // Reset the random number generator!
+        generator_.seed(gid_);
+    }
+
+    /// Advance the Poisson generator until end of epoch
+    /// If a rate change is occurred during this ep the new rate will be used
+    /// including possible interpolation between steps.
+    void advance(epoch ep, std::vector<spike>& spikes)
+    {
+        // Convert the epoch end to step
+        unsigned long epoch_end_step = time_to_step_idx(ep.tfinal, sample_delta_);
+
+        // If the start time of the neuron is after the end of the epoch
+        if (start_step_ > epoch_end_step) {
+            return;
+        }
+
+        auto end_step = std::min(stop_step_, epoch_end_step);
+
+        // epoch is after end of cell
+        if (current_step_ >= end_step) {
+            return;
+        }
+
+        // We are in an active epoch, start stepping!
+        do {
+            // Should we change the rates in the current step?
+
+            if (current_step_ == next_rates_it_->first) {
+                // update the itterators
+                current_rate_it_++;
+                next_rates_it_++;
+                update_rates();
+            }
+
+            // We are working with small probabilities use double
+            // rate = base rate + nr steps since start rate * delta per step
+            double spike_rate = base_rate_ + (current_step_ - current_rate_it_->first) * rate_delta_step_ ;
+
+            // This dice roll takes 93% of the runtime of this cell.
+            auto dice_roll = distribution_(generator_);
+            if (dice_roll < spike_rate) {
+                spikes.push_back({ { gid_, 0 }, current_step_ * sample_delta_ });
+            }
+
+        } while (++current_step_ < end_step);
+
+    }
+
+    // Small helper function for 'rounding' up to whole multiples of dt
+    // public for testing
+    static  unsigned long time_to_step_idx(time_type time, time_type dt) {
+        unsigned long whole_multiples = static_cast<unsigned long>(time / dt);
+
+        if (whole_multiples * dt < time) {
+            return whole_multiples + 1;
+        }
+
+        return whole_multiples;
+    }
+
+private:
+    /// We need to do some book keeping when we have a rate_change step
+    ///   - Calculate the base rate and the delta per step
+    void update_rates() {
+        // First calculate the rate per time step (if we interpolate)
+        auto steps_this_rate = next_rates_it_->first - current_rate_it_->first;
```

Comment:
I'm not sure that the compiler is smart enough to recognize that if `! interpolate_`, this all collapses down to:
```
rate_delta_step_ = 0;
base_rate_ = current_rate_it_->second
```

So maybe an `if else` is needed here instead.